File: Recursion/permutations/permutations.c
/******************************************************************************
 *
 *  Print all permutations of length n
 *
 *****************************************************************************/
#include <stdio.h>
#include <stdlib.h>

void printPermutationsRecursion(int n, int *array, int i) {
    if (i == n) {
        for(int j = 0; j < n; ++j)
            printf("%d ", array[j]);
        printf("\n");
        return;
    }

    for(int j = i; j < n; ++j) {
        int tmp = array[j];
        array[j] = array[i];
        array[i] = tmp;
        printPermutationsRecursion(n, array, i + 1);
        array[i] = array[j];
        array[j] = tmp;
    } 
}

void printPermutations(int n) {
    int *array = malloc(n * sizeof(int));
    for(int i = 0; i < n; ++i)
        array[i] = i + 1;
    printPermutationsRecursion(n, array, 0);
    free(array);
}

int main(int argc, char **argv) {
    printPermutations(atoi(argv[1]));
    return 0;
} 

  

File: Recursion/permutations/Permutations.java
/******************************************************************************
 * Print all permutations of an array of length n
 *****************************************************************************/
public class Permutations {
    public static void main(String[] args) {
        printPermutations(args);
    }

    public static void printPermutations(Object[] array) {
        printPermutations(array, 0);
    }

    private static void printPermutations(Object[] array, int i) {
        if (i == array.length) {
            for (Object obj : array) System.out.print(" " + obj);
            System.out.println();
            return;
        }

        for (int j = i; j < array.length; ++j) {
            swap(array, i, j);
            printPermutations(array, i + 1);
            swap(array, i, j);
        }
    }

    private static void swap(Object[] array, int i, int j) {
        Object tmp = array[j];
        array[j] = array[i];
        array[i] = tmp;
    }
}
